### DISCUSSION TITLE
2D CAD - shapes are fused on export

### DISCUSSION THREAD

**Discussion by WayneSherman** [gumyr/build123d#699] [Q&A] [answered]
I am testing out build123d for 2D CAD with export to DXF and SVG.
(using [CQ-Editor 0.30-dev-5](https://github.com/jdegenstein/jmwright-CQ-Editor/releases/tag/0.30-dev-5) , based on gumyr/build123d from Mar 31, 2024)

My first (very simple) test is drawing a floor plan of a rectangular building with three rooms. 

```
# NOTE: Drawing units are feet

#         225 ft bldg length
# ---------------------------------
# |   60   |   60   |     105     |
# |   x    |   x    |      x      |  50 ft bldg width
# |   50   |   50   |     50      |
# ---------------------------------
#   room1    room2      room3
```

For this first test I am drawing three rectangles and exporting the sketch.  But in the exported drawings, **_the rectangles fuse together into a single rectangle_** instead of 3 separate rectangles.  Following the patterns in the example docs I wrote the following code:
```python
from build123d import *

bldg_width, bldg_length = [50, 225]
room1_width, room2_width = [60, 60]
room3_width = bldg_length - room1_width - room2_width
room_height = bldg_width

with BuildSketch() as MySketch:
    Rectangle(room1_width, room_height, align=(Align.MIN, Align.MIN))
    with Locations((room1_width,0)):
        Rectangle(room2_width, room_height, align=(Align.MIN, Align.MIN))
    with Locations((room1_width+room2_width,0)):
        Rectangle(room3_width, room_height, align=(Align.MIN, Align.MIN))

exporter = ExportDXF(unit=Unit.FT, line_weight=0.5)
exporter.add_layer("Layer 1")
exporter.add_shape(MySketch.sketch, layer="Layer 1")
exporter.write("Simple_Building.dxf")

# Since the real life drawing extents are too large for inkscape in this case,
# we scale the sketch so the SVG drawing scale will be 1/16in = 1ft
ScaledSketch = scale(MySketch.sketch, by=1/16)
exporter = ExportSVG(unit=Unit.IN, line_weight=0.2)
exporter.add_layer("Layer 1")
exporter.add_shape(ScaledSketch, layer="Layer 1")
exporter.write("Simple_Building.svg")

show_object(MySketch)  #syntax to show the sketch in CQ-Editor 
```

Thanks for any help.

**Comment by WayneSherman** [comment]
I found a way (in [discussion](https://github.com/gumyr/build123d/discussions/260#discussioncomment-6315950)) by using "Algebra Mode" and creating a Sketch with "children":

```python
room1 = Rectangle(room1_width, room_height, align=Align.MIN)
room2 = Pos(room1_width, 0) * Rectangle(room2_width, room_height, align=Align.MIN)
room3 = Pos(room1_width + room2_width, 0) * Rectangle(room3_width, room_height, align=Align.MIN)

#MySketch = room1 + room2 + room3  # this adds the shapes using Mode.ADD which fuses them together
MySketch = Sketch(children=[room1, room2, room3])  #this adds the shapes to a sketch such that they stay separate
```

Can the same thing be accomplished in "Builder Mode"?

**Reply by jdegenstein** [reply, marked as answer]
@WayneSherman your approach in algebra mode is to essentially create a `Compound` by using `Sketch` (in build123d `Sketch` is a type of `Compound` also known as an assembly). In general, the builders are designed to by default automatically fuse objects together, so what you observed is expected behavior (n.b. see the `mode` parameters of the objects e.g. `Rectangle` in this case and also of the builders). If the objects "don't touch" then they will remain as separate shapes, but they will be (by default) automatically "merged" otherwise. When talking about `BuildPart`, I often say that it is intended for "single solid design" but it can be used for multiple solids subject to the above constraints.

Back to your question -- can you do this in builder mode? The answer is yes, by combining a bit of both approaches:
```py
with BuildSketch() as room1:
    Rectangle(room1_width, room_height, align=(Align.MIN, Align.MIN))

with BuildSketch() as room2:
    with Locations((room1_width,0)):
        Rectangle(room2_width, room_height, align=(Align.MIN, Align.MIN))

with BuildSketch() as room3:
    with Locations((room1_width+room2_width,0)):
        Rectangle(room3_width, room_height, align=(Align.MIN, Align.MIN))

face_assembly = Compound()
face_assembly.children = [room1.sketch, room2.sketch, room3.sketch]
```
Obviously algebra mode is much more space efficient in this simple case, but let's assume that your room designs get more complex -- the "overhead" of using builder mode would be less significant. I personally appreciate the visually hierarchical (i.e. indented) structure of builder mode as well.

**Reply by WayneSherman** [reply]
Thank you, that works.

Using Compound with children seems to break the ability to "scale":

```python
scale(face_assembly, by=1/16)  # this is not scaling the Compound.children objects
```
EDIT:  I was using the scale function incorrectly.  It doesn't operate on the passed in object directly, but creates a copy that is returned.  It works like this:
```python
face_assembly = scale(face_assembly, by=1/16)
```
